politeiad/cmd/legacypoliteia/import.go

```diff
@@ -0,0 +1,335 @@
+// Copyright (c) 2021 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package main
+
+import (
+	"encoding/base64"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	"github.com/decred/politeia/politeiad/api/v1/mime"
+	backend "github.com/decred/politeia/politeiad/backendv2"
+	"github.com/decred/politeia/politeiad/backendv2/tstorebe/store"
+	"github.com/decred/politeia/politeiad/plugins/comments"
+	"github.com/decred/politeia/politeiad/plugins/ticketvote"
+	"github.com/decred/politeia/politeiad/plugins/usermd"
+	"github.com/decred/politeia/util"
+	"github.com/pkg/errors"
+)
+
+const (
+	dataDescriptorCastVoteDetails = ticketvote.PluginID + "-castvote-v1"
+	dataDescriptorVoteCollider    = ticketvote.PluginID + "-vcollider-v1"
+	dataDescriptorAuthDetails     = ticketvote.PluginID + "-auth-v1"
+	dataDescriptorVoteDetails     = ticketvote.PluginID + "-vote-v1"
+
+	dataDescriptorCommentAdd  = comments.PluginID + "-add-v1"
+	dataDescriptorCommentDel  = comments.PluginID + "-del-v1"
+	dataDescriptorCommentVote = comments.PluginID + "-vote-v1"
+)
+
+// importProposal imports the specified legacy proposal into tstore.
+//
+// This function assumes that the proposal does not yet exist in tstore.
+// Handling proposals that have been partially added is done by the
+// fsckProposal() function.
+//
+// The steps this function executes to accomplish the import are the following:
+//
+// 1. Create a new tlog tree for the legacy proposal.
+//
+// 2. Replace the git backend tokens with the tstore backend token in the
+//    following structures:
+//      - Record Metadata
+//      - Vote Details Params
+//      - Vote Metadata LinkTo
+//
+// 3. Import auth details blob.
+//
+// 4. Import vote details blob.
+//
+// 5. Import comment blobs and vote blobs concurrently. This is the most costly
+//    process of importing, and is why it is done concurrently.
+//
+// 6. Add git token to tstore token mapping to the memory cache for all RFP
+// 	  parent proposals. The parent token of all RFP submissions are updated
+//    with the tstore RFP parent token.
+func importProposal(legacyDir, gitToken string, cmd *importCmd) error {
+	// Read and unmarshal proposal from json file.
+	prop, err := loadProposal(legacyDir, gitToken)
+	if err != nil {
+		return err
+	}
+
+	// Create a new tlog tree for the legacy record.
+	token, err := cmd.tstore.RecordNew()
+	if err != nil {
+		return err
+	}
+
+	// Replace gitbe tokens with the new tstorebe token.
+	err = updateProposalWithTstoreToken(prop, hex.EncodeToString(token), cmd)
+	if err != nil {
+		return err
+	}
+
```

Comment:
Any updates to the proposal should be done here.  I don't have to have state changes nested inside of functions.  Doing it all in the same place makes it easier to audit. This function should read as:

1. Load the proposal
2. Create the tstore record
3. Update the proposal with the tstore token
4. Import the different parts of the proposal into tstore

Reply:
If this is not possible for all scenerios then we should talk it through.